src/python/grpcio/grpc/experimental/aio/_interceptor.py

```diff
@@ -325,14 +308,193 @@ async def debug_error_string(self) -> Optional[str]:
+                    return call_or_response
+                else:
+                    return UnaryUnaryCallResponse(call_or_response)
+
+            else:
+                return UnaryUnaryCall(
+                    request, _timeout_to_deadline(client_call_details.timeout),
+                    client_call_details.metadata,
+                    client_call_details.credentials,
+                    client_call_details.wait_for_ready, self._channel,
+                    client_call_details.method, request_serializer,
+                    response_deserializer, self._loop)
+
+        client_call_details = ClientCallDetails(method, timeout, metadata,
+                                                credentials, wait_for_ready)
+        return await _run_interceptor(iter(interceptors), client_call_details,
+                                      request)
+
     def __await__(self):
         call = yield from self._interceptors_task.__await__()
         response = yield from call.__await__()
         return response
 
-    async def wait_for_connection(self) -> None:
-        call = await self._interceptors_task
-        return await call.wait_for_connection()
+
+class CallResponseIterator:
+
+    _intercepted_call: UnaryStreamCall
+
+    def __init__(self) -> None:
+        self._intercepted_call = None
+
+    def set_call(self, call: UnaryStreamCall) -> None:
+        self._intercepted_call = call
+
+    async def _forward_responses(self) -> ResponseType:
+        assert self._intercepted_call, "Before start iterating set_call must be called"
+
+        async for response in self._intercepted_call:
+            yield response
+
+    def __aiter__(self) -> AsyncIterable[ResponseType]:
+        return self._forward_responses()
+
+
+class InterceptedUnaryStreamCall(InterceptedCall, _base_call.UnaryStreamCall):
+    """Used for running a `UnaryStreamCall` wrapped by interceptors."""
+
+    _loop: asyncio.AbstractEventLoop
+    _channel: cygrpc.AioChannel
+    _call_response_iterator: CallResponseIterator
+    _response_aiter: AsyncIterable[ResponseType]
+    _intercepted_response_iterator: Optional[AsyncIterable[ResponseType]]
+
+    # pylint: disable=too-many-arguments
+    def __init__(self, interceptors: Sequence[UnaryStreamClientInterceptor],
+                 request: RequestType, timeout: Optional[float],
+                 metadata: MetadataType,
+                 credentials: Optional[grpc.CallCredentials],
+                 wait_for_ready: Optional[bool], channel: cygrpc.AioChannel,
+                 method: bytes, request_serializer: SerializingFunction,
+                 response_deserializer: DeserializingFunction,
+                 loop: asyncio.AbstractEventLoop) -> None:
+        self._loop = loop
+        self._channel = channel
+        self._response_aiter = self._wait_for_interceptor_task_response_iterator()
+        self._call_response_iterator = CallResponseIterator()
+        self._intercepted_response_iterator = None
```

Comment:
Why we need to split `_intercepted_response_iterator` and `_response_aiter`?

Reply:
`_response_aiter ` its the instantiation of the response asynchronous iterator [1] provided by the class which can be used later on by the caller by iterating on the `InterceptedUnaryStreamCall` instance object or by using the `read` function. In both use cases, we don't want to restart - instantiate again - the interior but resume from the las consumed message.

`_intercepted_response_iterator` is what is finally provided by the interceptors, so it is used behind the scenes by the `_response_aiter`.